File: VB/App_Code/CardViewDataHelper.vb
﻿Imports DevExpress.Web
Imports System
Imports System.Collections.Specialized
Imports System.Data
Imports System.Web

Public Module CardViewDataHelper
	Public ReadOnly Property Data() As DataTable
		Get
			If HttpContext.Current.Session("data") Is Nothing Then
				Dim table = New DataTable()
				table.Columns.Add("ID", GetType(Integer))
				table.Columns.Add("Data", GetType(String))
				table.PrimaryKey = New DataColumn() { table.Columns("ID") }
				For i As Integer = 0 To 9
					table.Rows.Add(New Object() { i, "row " & i.ToString() })
				Next i
				HttpContext.Current.Session("data") = table
			End If
			Return CType(HttpContext.Current.Session("data"), DataTable)
		End Get
	End Property
	Public Sub InsertCard(ByVal sender As Object, ByVal e As DevExpress.Web.Data.ASPxDataInsertingEventArgs)
		InsertRow(e.NewValues)
		BindCardView(TryCast(sender, ASPxCardView))
		e.Cancel = True
	End Sub
	Public Sub UpdateCard(ByVal sender As Object, ByVal e As DevExpress.Web.Data.ASPxDataUpdatingEventArgs)
		UpdateRow(e.NewValues, CInt(Math.Truncate(e.Keys(0))))
		BindCardView(TryCast(sender, ASPxCardView))
		e.Cancel = True
	End Sub
	Public Sub DeleteCard(ByVal sender As Object, ByVal e As DevExpress.Web.Data.ASPxDataDeletingEventArgs)
		DeleteRow(CInt(Math.Truncate(e.Keys(0))))
		BindCardView(TryCast(sender, ASPxCardView))
		e.Cancel = True
	End Sub
	Private Sub BindCardView(ByVal cv As ASPxCardView)
		cv.CancelEdit()
		cv.DataSource = Data
		cv.DataBind()
	End Sub
	Private Function GetNewId(ByVal table As DataTable) As Integer
		If table.Rows.Count = 0 Then
			Return 0
		End If
		Dim max As Integer = Convert.ToInt32(table.Rows(0)("ID"))
		For i As Integer = 1 To table.Rows.Count - 1
			If Convert.ToInt32(table.Rows(i)("ID")) > max Then
				max = Convert.ToInt32(table.Rows(i)("ID"))
			End If
		Next i
		Return max + 1
	End Function
	Private Sub UpdateRow(ByVal newValues As OrderedDictionary, ByVal ID As Integer)
		Dim row = Data.Rows.Find(ID)
		row("Data") = newValues("Data")

	End Sub
	Private Sub InsertRow(ByVal newValues As OrderedDictionary)
		Dim row As DataRow = Data.NewRow()
		row("ID") = GetNewId(Data)
		row("Data") = newValues("Data")
		Data.Rows.Add(row)
	End Sub
	Private Sub DeleteRow(ByVal ID As Integer)
		Dim row = Data.Rows.Find(ID)
		Data.Rows.Remove(row)
	End Sub
End Module

File: CS/App_Code/CardViewDataHelper.cs
using DevExpress.Web;
using System;
using System.Collections.Specialized;
using System.Data;
using System.Web;

public static class CardViewDataHelper {
	public static DataTable Data {
		get {
			if(HttpContext.Current.Session["data"] == null) {
				var table = new DataTable();
				table.Columns.Add("ID", typeof(int));
				table.Columns.Add("Data", typeof(string));
				table.PrimaryKey = new DataColumn[] { table.Columns["ID"] };
				for(int i = 0; i < 10; i++) {
					table.Rows.Add(new object[] { i, "row " + i.ToString() });
				}
				HttpContext.Current.Session["data"] = table;
			}
			return (DataTable)HttpContext.Current.Session["data"];
		}
	}
	public static void InsertCard(object sender, DevExpress.Web.Data.ASPxDataInsertingEventArgs e) {
		InsertRow(e.NewValues);
		BindCardView(sender as ASPxCardView);
		e.Cancel = true;
	}
	public static void UpdateCard(object sender, DevExpress.Web.Data.ASPxDataUpdatingEventArgs e) {
		UpdateRow(e.NewValues, (int)e.Keys[0]);
		BindCardView(sender as ASPxCardView);
		e.Cancel = true;
	}
	public static void DeleteCard(object sender, DevExpress.Web.Data.ASPxDataDeletingEventArgs e) {
		DeleteRow((int)e.Keys[0]);
		BindCardView(sender as ASPxCardView);
		e.Cancel = true;
	}
	static void BindCardView(ASPxCardView cv) {
		cv.CancelEdit();
		cv.DataSource = Data;
		cv.DataBind();
	}
	static int GetNewId(DataTable table) {
		if(table.Rows.Count == 0) return 0;
		int max = Convert.ToInt32(table.Rows[0]["ID"]);
		for(int i = 1; i < table.Rows.Count; i++) {
			if(Convert.ToInt32(table.Rows[i]["ID"]) > max)
				max = Convert.ToInt32(table.Rows[i]["ID"]);
		}
		return max + 1;
	}
	static void UpdateRow(OrderedDictionary newValues, int ID) {
		var row = Data.Rows.Find(ID);
		row["Data"] = newValues["Data"];
		
	}
	static void InsertRow(OrderedDictionary newValues) {
		DataRow row = Data.NewRow();
		row["ID"] = GetNewId(Data);
		row["Data"] = newValues["Data"];
		Data.Rows.Add(row);
	}
	static void DeleteRow(int ID) {
		var row = Data.Rows.Find(ID);
		Data.Rows.Remove(row);
	}
}
